File: aws/developing-web-apps-amplify/code/terraform/lambda-function.js
// Lambda function for Todo API backend
// This function handles CRUD operations for the Todo application

const AWS = require('aws-sdk');
const { v4: uuidv4 } = require('uuid');

// Initialize DynamoDB DocumentClient
const dynamodb = new AWS.DynamoDB.DocumentClient();

// Get table name from environment variable
const TABLE_NAME = process.env.DYNAMODB_TABLE_NAME || '${table_name}';

// CORS headers for all responses
const CORS_HEADERS = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': process.env.CORS_ORIGIN || '*',
    'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token',
    'Access-Control-Allow-Methods': 'GET,POST,PUT,DELETE,OPTIONS'
};

/**
 * Main Lambda handler function
 * @param {Object} event - API Gateway event object
 * @param {Object} context - Lambda context object
 * @returns {Object} - API Gateway response object
 */
exports.handler = async (event, context) => {
    console.log('Event:', JSON.stringify(event, null, 2));
    console.log('Context:', JSON.stringify(context, null, 2));

    try {
        // Extract HTTP method and path from event
        const { httpMethod, path, pathParameters, body, queryStringParameters } = event;
        
        // Route requests based on HTTP method and path
        switch (httpMethod) {
            case 'GET':
                if (pathParameters && pathParameters.id) {
                    return await getTodoById(pathParameters.id);
                } else {
                    return await getAllTodos(queryStringParameters);
                }
                
            case 'POST':
                return await createTodo(JSON.parse(body || '{}'));
                
            case 'PUT':
                if (pathParameters && pathParameters.id) {
                    return await updateTodo(pathParameters.id, JSON.parse(body || '{}'));
                } else {
                    return createErrorResponse(400, 'Todo ID is required for PUT requests');
                }
                
            case 'DELETE':
                if (pathParameters && pathParameters.id) {
                    return await deleteTodo(pathParameters.id);
                } else {
                    return createErrorResponse(400, 'Todo ID is required for DELETE requests');
                }
                
            case 'OPTIONS':
                return createResponse(200, { message: 'CORS preflight' });
                
            default:
                return createErrorResponse(405, `Method ${httpMethod} not allowed`);
        }
    } catch (error) {
        console.error('Error processing request:', error);
        return createErrorResponse(500, 'Internal server error', error.message);
    }
};

/**
 * Get all todos with optional filtering and pagination
 * @param {Object} queryParams - Query string parameters
 * @returns {Object} - API response with todos list
 */
async function getAllTodos(queryParams = {}) {
    try {
        console.log('Getting all todos with params:', queryParams);
        
        const params = {
            TableName: TABLE_NAME
        };
        
        // Add pagination if provided
        if (queryParams && queryParams.lastKey) {
            params.ExclusiveStartKey = JSON.parse(decodeURIComponent(queryParams.lastKey));
        }
        
        // Add limit if provided
        if (queryParams && queryParams.limit) {
            params.Limit = parseInt(queryParams.limit);
        }
        
        const result = await dynamodb.scan(params).promise();
        
        // Sort todos by creation date (newest first)
        const sortedTodos = result.Items.sort((a, b) => 
            new Date(b.createdAt) - new Date(a.createdAt)
        );
        
        const response = {
            todos: sortedTodos,
            count: sortedTodos.length,
            totalCount: result.Count
        };
        
        // Include pagination token if there are more items
        if (result.LastEvaluatedKey) {
            response.nextKey = encodeURIComponent(JSON.stringify(result.LastEvaluatedKey));
        }
        
        console.log(`Successfully retrieved ${sortedTodos.length} todos`);
        return createResponse(200, response);
        
    } catch (error) {
        console.error('Error getting todos:', error);
        return createErrorResponse(500, 'Failed to retrieve todos', error.message);
    }
}

/**
 * Get a specific todo by ID
 * @param {string} todoId - Todo ID
 * @returns {Object} - API response with todo item
 */
async function getTodoById(todoId) {
    try {
        console.log('Getting todo by ID:', todoId);
        
        const params = {
            TableName: TABLE_NAME,
            Key: { id: todoId }
        };
        
        const result = await dynamodb.get(params).promise();
        
        if (!result.Item) {
            return createErrorResponse(404, 'Todo not found');
        }
        
        console.log('Successfully retrieved todo:', result.Item.id);
        return createResponse(200, result.Item);
        
    } catch (error) {
        console.error('Error getting todo by ID:', error);
        return createErrorResponse(500, 'Failed to retrieve todo', error.message);
    }
}

/**
 * Create a new todo item
 * @param {Object} todoData - Todo data from request body
 * @returns {Object} - API response with created todo
 */
async function createTodo(todoData) {
    try {
        console.log('Creating new todo:', todoData);
        
        // Validate required fields
        if (!todoData.title || typeof todoData.title !== 'string') {
            return createErrorResponse(400, 'Title is required and must be a string');
        }
        
        // Create new todo item with generated ID and timestamp
        const todo = {
            id: uuidv4(),
            title: todoData.title.trim(),
            completed: Boolean(todoData.completed),
            createdAt: new Date().toISOString(),
            updatedAt: new Date().toISOString(),
            // Add optional fields if provided
            ...(todoData.description && { description: todoData.description.trim() }),
            ...(todoData.priority && { priority: todoData.priority }),
            ...(todoData.dueDate && { dueDate: todoData.dueDate })
        };
        
        const params = {
            TableName: TABLE_NAME,
            Item: todo,
            // Ensure we don't overwrite existing items
            ConditionExpression: 'attribute_not_exists(id)'
        };
        
        await dynamodb.put(params).promise();
        
        console.log('Successfully created todo:', todo.id);
        return createResponse(201, todo);
        
    } catch (error) {
        console.error('Error creating todo:', error);
        if (error.code === 'ConditionalCheckFailedException') {
            return createErrorResponse(409, 'Todo with this ID already exists');
        }
        return createErrorResponse(500, 'Failed to create todo', error.message);
    }
}

/**
 * Update an existing todo item
 * @param {string} todoId - Todo ID to update
 * @param {Object} updateData - Updated todo data
 * @returns {Object} - API response with updated todo
 */
async function updateTodo(todoId, updateData) {
    try {
        console.log('Updating todo:', todoId, updateData);
        
        // First, check if the todo exists
        const getParams = {
            TableName: TABLE_NAME,
            Key: { id: todoId }
        };
        
        const existingTodo = await dynamodb.get(getParams).promise();
        if (!existingTodo.Item) {
            return createErrorResponse(404, 'Todo not found');
        }
        
        // Build update expression dynamically
        const updateExpressions = [];
        const expressionAttributeNames = {};
        const expressionAttributeValues = {};
        
        // Update allowed fields
        const allowedFields = ['title', 'completed', 'description', 'priority', 'dueDate'];
        
        allowedFields.forEach(field => {
            if (updateData.hasOwnProperty(field)) {
                const attrName = `#${field}`;
                const attrValue = `:${field}`;
                
                updateExpressions.push(`${attrName} = ${attrValue}`);
                expressionAttributeNames[attrName] = field;
                
                // Validate and process field values
                switch (field) {
                    case 'title':
                        if (typeof updateData[field] !== 'string' || !updateData[field].trim()) {
                            return createErrorResponse(400, 'Title must be a non-empty string');
                        }
                        expressionAttributeValues[attrValue] = updateData[field].trim();
                        break;
                    case 'completed':
                        expressionAttributeValues[attrValue] = Boolean(updateData[field]);
                        break;
                    default:
                        expressionAttributeValues[attrValue] = updateData[field];
                }
            }
        });
        
        // Always update the updatedAt timestamp
        updateExpressions.push('#updatedAt = :updatedAt');
        expressionAttributeNames['#updatedAt'] = 'updatedAt';
        expressionAttributeValues[':updatedAt'] = new Date().toISOString();
        
        if (updateExpressions.length === 1) { // Only updatedAt
            return createErrorResponse(400, 'No valid fields to update');
        }
        
        const updateParams = {
            TableName: TABLE_NAME,
            Key: { id: todoId },
            UpdateExpression: `SET ${updateExpressions.join(', ')}`,
            ExpressionAttributeNames: expressionAttributeNames,
            ExpressionAttributeValues: expressionAttributeValues,
            ReturnValues: 'ALL_NEW'
        };
        
        const result = await dynamodb.update(updateParams).promise();
        
        console.log('Successfully updated todo:', todoId);
        return createResponse(200, result.Attributes);
        
    } catch (error) {
        console.error('Error updating todo:', error);
        return createErrorResponse(500, 'Failed to update todo', error.message);
    }
}

/**
 * Delete a todo item
 * @param {string} todoId - Todo ID to delete
 * @returns {Object} - API response confirming deletion
 */
async function deleteTodo(todoId) {
    try {
        console.log('Deleting todo:', todoId);
        
        const params = {
            TableName: TABLE_NAME,
            Key: { id: todoId },
            // Return the deleted item to confirm it existed
            ReturnValues: 'ALL_OLD'
        };
        
        const result = await dynamodb.delete(params).promise();
        
        if (!result.Attributes) {
            return createErrorResponse(404, 'Todo not found');
        }
        
        console.log('Successfully deleted todo:', todoId);
        return createResponse(200, { 
            message: 'Todo deleted successfully',
            deletedTodo: result.Attributes
        });
        
    } catch (error) {
        console.error('Error deleting todo:', error);
        return createErrorResponse(500, 'Failed to delete todo', error.message);
    }
}

/**
 * Create a successful API Gateway response
 * @param {number} statusCode - HTTP status code
 * @param {Object} data - Response data
 * @returns {Object} - Formatted API Gateway response
 */
function createResponse(statusCode, data) {
    return {
        statusCode,
        headers: CORS_HEADERS,
        body: JSON.stringify(data)
    };
}

/**
 * Create an error API Gateway response
 * @param {number} statusCode - HTTP status code
 * @param {string} message - Error message
 * @param {string} details - Optional error details
 * @returns {Object} - Formatted API Gateway error response
 */
function createErrorResponse(statusCode, message, details = null) {
    const errorResponse = {
        error: message,
        statusCode,
        timestamp: new Date().toISOString()
    };
    
    if (details) {
        errorResponse.details = details;
    }
    
    console.error('API Error:', errorResponse);
    
    return {
        statusCode,
        headers: CORS_HEADERS,
        body: JSON.stringify(errorResponse)
    };
}

File: aws/building-global-active-active-architecture/code/terraform/lambda_function.py
"""
Multi-region active-active application handler
Supports CRUD operations with automatic regional optimization
"""

import json
import boto3
import time
import os
from decimal import Decimal

# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb')
table_name = os.environ.get('TABLE_NAME', '${table_name}')
table = dynamodb.Table(table_name)

def lambda_handler(event, context):
    """
    Multi-region active-active application handler
    Supports CRUD operations with automatic regional optimization
    """
    
    try:
        # Get HTTP method and path
        method = event.get('httpMethod', 'GET')
        path = event.get('path', '/')
        
        # Parse request body for POST/PUT requests
        body = {}
        if event.get('body'):
            body = json.loads(event['body'])
        
        # Get AWS region from context
        region = context.invoked_function_arn.split(':')[3]
        
        # Route based on HTTP method and path
        if method == 'GET' and path == '/health':
            return health_check(region)
        elif method == 'GET' and path.startswith('/user/'):
            user_id = path.split('/')[-1]
            return get_user_data(user_id, region)
        elif method == 'POST' and path == '/user':
            return create_user_data(body, region)
        elif method == 'PUT' and path.startswith('/user/'):
            user_id = path.split('/')[-1]
            return update_user_data(user_id, body, region)
        elif method == 'GET' and path == '/users':
            return list_users(region)
        else:
            return {
                'statusCode': 404,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Methods': 'GET,POST,PUT,DELETE,OPTIONS',
                    'Access-Control-Allow-Headers': 'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'
                },
                'body': json.dumps({'error': 'Not found'})
            }
            
    except Exception as e:
        print(f"Error processing request: {str(e)}")
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'error': str(e), 
                'region': region,
                'table_name': table_name
            })
        }

def health_check(region):
    """Health check endpoint for load balancer"""
    try:
        # Test DynamoDB connectivity
        table.scan(Limit=1)
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'status': 'healthy',
                'region': region,
                'timestamp': int(time.time()),
                'table_name': table_name,
                'service': 'multi-region-active-active-app'
            })
        }
    except Exception as e:
        print(f"Health check failed: {str(e)}")
        return {
            'statusCode': 503,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'status': 'unhealthy',
                'region': region,
                'error': str(e),
                'timestamp': int(time.time())
            })
        }

def get_user_data(user_id, region):
    """Get user data with regional context"""
    try:
        # Get latest record for user
        response = table.query(
            KeyConditionExpression='userId = :userId',
            ExpressionAttributeValues={':userId': user_id},
            ScanIndexForward=False,
            Limit=1
        )
        
        if response['Items']:
            item = response['Items'][0]
            # Convert Decimal to float for JSON serialization
            item = json.loads(json.dumps(item, default=decimal_default))
            return {
                'statusCode': 200,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({
                    'user': item,
                    'served_from_region': region,
                    'table_name': table_name
                })
            }
        else:
            return {
                'statusCode': 404,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({
                    'error': 'User not found',
                    'userId': user_id,
                    'served_from_region': region
                })
            }
            
    except Exception as e:
        print(f"Error getting user data: {str(e)}")
        raise Exception(f"Error getting user data: {str(e)}")

def create_user_data(body, region):
    """Create new user data with regional tracking"""
    try:
        user_id = body.get('userId')
        if not user_id:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({
                    'error': 'userId is required',
                    'served_from_region': region
                })
            }
        
        timestamp = int(time.time() * 1000)  # milliseconds for better precision
        
        item = {
            'userId': user_id,
            'timestamp': timestamp,
            'data': body.get('data', {}),
            'created_region': region,
            'last_updated': timestamp,
            'version': 1,
            'table_name': table_name
        }
        
        table.put_item(Item=item)
        
        return {
            'statusCode': 201,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'message': 'User created successfully',
                'userId': user_id,
                'created_in_region': region,
                'timestamp': timestamp,
                'table_name': table_name
            })
        }
        
    except Exception as e:
        print(f"Error creating user data: {str(e)}")
        raise Exception(f"Error creating user data: {str(e)}")

def update_user_data(user_id, body, region):
    """Update user data with conflict resolution"""
    try:
        timestamp = int(time.time() * 1000)
        
        # Create new record for the update (time-series approach)
        item = {
            'userId': user_id,
            'timestamp': timestamp,
            'data': body.get('data', {}),
            'updated_region': region,
            'last_updated': timestamp,
            'version': body.get('version', 1) + 1,
            'table_name': table_name
        }
        
        table.put_item(Item=item)
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'message': 'User updated successfully',
                'userId': user_id,
                'updated_in_region': region,
                'timestamp': timestamp,
                'table_name': table_name
            })
        }
        
    except Exception as e:
        print(f"Error updating user data: {str(e)}")
        raise Exception(f"Error updating user data: {str(e)}")

def list_users(region):
    """List users with pagination support"""
    try:
        # Simple scan with limit (in production, use GSI for better performance)
        response = table.scan(Limit=20)
        
        users = []
        processed_users = set()
        
        for item in response['Items']:
            user_id = item['userId']
            if user_id not in processed_users:
                users.append({
                    'userId': user_id,
                    'last_updated': int(item['last_updated']),
                    'version': int(item.get('version', 1)),
                    'created_region': item.get('created_region', 'unknown'),
                    'latest_region': item.get('updated_region', item.get('created_region', 'unknown'))
                })
                processed_users.add(user_id)
        
        # Sort by last_updated descending
        users.sort(key=lambda x: x['last_updated'], reverse=True)
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'users': users,
                'count': len(users),
                'served_from_region': region,
                'table_name': table_name
            })
        }
        
    except Exception as e:
        print(f"Error listing users: {str(e)}")
        raise Exception(f"Error listing users: {str(e)}")

def decimal_default(obj):
    """JSON serializer for objects not serializable by default json code"""
    if isinstance(obj, Decimal):
        return float(obj)
    raise TypeError(f"Object of type {type(obj)} is not JSON serializable")